src/storage/src/source/persist_source.rs

```diff
@@ -190,3 +194,218 @@ where
+            .lock()
+            .await
+            .open(persist_location_moveable)
+            .await
+            .expect("could not open persist client")
+            .clone();
+
+        let mut read = persist_client
+            .open_reader::<SourceData, (), mz_repr::Timestamp, mz_repr::Diff>(data_shard.clone())
+            .await
+            .expect("could not open persist shard");
+
+        // Aggressively downgrade `since`, to not hold back compaction.
+        read.downgrade_since(as_of_moveable.clone()).await;
+
+        let snapshots_hollow_batches = read
+            .snapshot_hollow_batch(as_of_moveable.clone())
+            .await
+            .expect("cannot serve requested as_of");
+
+        // First, yield all the updates from the snapshot.
+        for batch in snapshots_hollow_batches {
+            yield batch;
+        }
+
+        // Then, listen continously and yield any new updates. This loop is expected to never
+        // finish.
+        let mut listen = read
+            .listen(as_of_moveable)
+            .await
+            .expect("cannot serve requested as_of");
+
+        loop {
+            yield listen.listen_hollow_batch().await;
+        }
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let (inner, token) =
+        crate::source::util::source(scope, "persist_source".to_string(), move |info| {
+            let waker_activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+            let waker = futures_util::task::waker(waker_activator);
+            let activator = scope.activator_for(&info.address[..]);
+
+            // There is a bit of a mismatch: listening on a ReadHandle will give us an Antichain<T>
+            // as a frontier in `Progress` messages while a timely source usually only has a single
+            // `Capability` that it can downgrade. We can work around that by using a
+            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
+            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
+            // we have to keep downgrading because we cannot simply drop the one that we don't
+            // need.
+
+            let mut current_ts = 0;
+
+            move |cap, output| {
+                let mut context = Context::from_waker(&waker);
+                // Bound execution of operator to prevent a single operator from hogging
+                // the CPU if there are many messages to process
+                let timer = Instant::now();
+
+                let mut i = 0;
+
+                while let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
+                    match item {
+                        Some(Ok(batch)) => {
+                            let session_cap = cap.delayed(&current_ts);
+                            let mut session = output.session(&session_cap);
+                            let progress = batch.generate_progress();
```

Comment:
I'd find this more straightforward to reason about if the stream item was something like `(EnrichedHollowBatch, Option<Antichain<T>>)` instead. maybe? curious what @aljoscha thinks here

it's moving a bit of logic out of persist and into the operators here, but the progress is a function of the stream of batches, not of an individual batch. said another way, if you swapped the order you emitted the batches above so that e.g. you interleaved snapshot and listen batches, the way this code is now would be incorrect and that wouldn't be super obvious. at least if we were doing the progress stuff above, it would be obvious that there was _some_ contract being upheld up there if someone went in and messed with the batch order

perhaps even better is to make a new `Subscription` struct in persist that looks a lot like Listen and is a thin wrapper for both a snapshot _and_ a Listen. then we can move all the frontier logic there

Reply:
Oh, that is great and so sensible. Will do.

Reply:
I put this in a new `Subscribe` struct; lmk if that was what you were thinking.

Reply:
Yeah, this is great!